joomla/joomla-cms
# Create Exception subclasses for the cache API, use them, check for connection failures in tests

#### Summary of Changes

Create subclasses of `RuntimeException` to be used for more specific fail conditions in the cache API and implement them:
- `JCacheExceptionConnecting` when an adapter tries to connect to an external service and fails
- `JCacheExceptionUnsupported` when the requested cache adapter is unsupported (either a class cannot be found or the `isSupported()` method checks return false

Removes the over-eager connection check in the Memcached adapter's `isSupported()` method.  The reasoning for this has been explained elsewhere.

Marks in the Redis adapter that it will throw Exceptions as of 4.0.

Makes the unit tests catch a connecting exception and skip the test if the platform can't be reached correctly.

In the Memcache(d) adapters null out the connection when throwing a connection exception; this will force the constructor to attempt the connection again if the class is instantiated after the first failure.
#### Testing Instructions

The cache API still functions as expected

On a connection failure for Memcache(d) and Redis a `JCacheExceptionConnecting` object (subclass of `RuntimeException` so no B/C break with existing try/catch code) is thrown

When a cache adapter is not supported a `JCacheExceptionUnsupported` object (subclass of `RuntimeException` so no B/C break with existing try/catch code) is thrown
#### Documentation Changes Required

N/A
